File: lib/features/player/data/sleep_timer.dart
import 'dart:async';

import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../settings/data/player_prefs.dart';
import '../../video/data/video_controller.dart';
import 'playback_controller.dart';

class SleepTimerState {
  const SleepTimerState({
    this.remaining,
    this.remainingTracks,
    this.waitingTrackEnd = false,
  });

  /// Countdown left in timed mode.
  final Duration? remaining;

  /// Track-count mode: stop once this many tracks finish (1 = current track).
  final int? remainingTracks;

  /// Countdown hit zero but "finish current track" is on — playback keeps
  /// going until the track ends.
  final bool waitingTrackEnd;

  bool get isActive =>
      remaining != null || remainingTracks != null || waitingTrackEnd;
}

/// Counts down and pauses playback, or stops after N tracks finish.
/// Tick-decremented rather than wall-clock-anchored on purpose: while audio
/// plays in the background iOS keeps the isolate alive (same as the
/// controller's 5s position timer), and once playback is paused and the app
/// suspends, a sleep timer has nothing left to do.
class SleepTimerController extends Notifier<SleepTimerState> {
  static const presetMinutes = [15, 30, 45, 60, 90];
  static const presetTrackCounts = [1, 2, 3, 5];
  static const fadeSeconds = 10;

  Timer? _tick;
  double? _fadeBaseVolume;
  bool _fadeOnVideo = false;

  @override
  SleepTimerState build() {
    ref.onDispose(() => _tick?.cancel());
    return const SleepTimerState();
  }

  void start(Duration duration) {
    _reset();
    state = SleepTimerState(remaining: duration);
    _tick = Timer.periodic(const Duration(seconds: 1), (_) => _onTick());
  }

  void stopAfterTracks(int count) {
    _reset();
    state = SleepTimerState(remainingTracks: count);
  }

  void cancel() {
    _reset();
    state = const SleepTimerState();
  }

  /// Called by the players when a track/video finishes; true means "stop
  /// now". In track-count mode intermediate completions just decrement.
  bool onTrackCompleted() {
    if (state.waitingTrackEnd) {
      state = const SleepTimerState();
      return true;
    }
    final tracks = state.remainingTracks;
    if (tracks == null) return false;
    if (tracks > 1) {
      state = SleepTimerState(remainingTracks: tracks - 1);
      return false;
    }
    state = const SleepTimerState();
    return true;
  }

  void _onTick() {
    final remaining = state.remaining;
    if (remaining == null) return;
    final next = remaining - const Duration(seconds: 1);
    if (next <= Duration.zero) {
      unawaited(_fire());
      return;
    }
    state = SleepTimerState(remaining: next);
    // No fade when the track is allowed to finish — nothing gets cut off.
    if (next.inSeconds <= fadeSeconds &&
        !ref.read(playerPrefsProvider).sleepFinishCurrentTrack) {
      final video = ref.read(videoControllerProvider).controller;
      if (_fadeBaseVolume == null) {
        _fadeOnVideo = video != null;
        _fadeBaseVolume = _fadeOnVideo
            ? video!.value.volume
            : ref.read(playbackControllerProvider.notifier).player.volume;
      }
      final volume = _fadeBaseVolume! * next.inSeconds / fadeSeconds;
      if (_fadeOnVideo) {
        video?.setVolume(volume);
      } else {
        ref.read(playbackControllerProvider.notifier).player.setVolume(volume);
      }
    }
  }

  Future<void> _fire() async {
    _tick?.cancel();
    _tick = null;
    if (ref.read(playerPrefsProvider).sleepFinishCurrentTrack && _isPlaying()) {
      state = const SleepTimerState(waitingTrackEnd: true);
      return;
    }
    state = const SleepTimerState();
    await ref.read(playbackControllerProvider.notifier).pause();
    if (ref.read(videoControllerProvider).hasVideo) {
      await ref.read(videoControllerProvider.notifier).pause();
    }
    _restoreVolume();
  }

  bool _isPlaying() {
    final video = ref.read(videoControllerProvider).controller;
    if (video != null) return video.value.isPlaying;
    return ref.read(playbackControllerProvider.notifier).player.playing;
  }

  void _reset() {
    _tick?.cancel();
    _tick = null;
    _restoreVolume();
  }

  void _restoreVolume() {
    final base = _fadeBaseVolume;
    _fadeBaseVolume = null;
    if (base == null) return;
    if (_fadeOnVideo) {
      ref.read(videoControllerProvider).controller?.setVolume(base);
    } else {
      ref.read(playbackControllerProvider.notifier).player.setVolume(base);
    }
  }
}

final sleepTimerProvider =
    NotifierProvider<SleepTimerController, SleepTimerState>(
      SleepTimerController.new,
    );

File: native/Tonari/Player/SleepTimer.swift
import Foundation

/// Whatever is playing in front: the audio player or the video player.
protocol SleepTarget: AnyObject {
    var volume: Float { get set }
    var isPlaying: Bool { get }
    func pause()
}

/// Pauses playback after a countdown or after a number of tracks. The last
/// ten seconds of a countdown fade out, unless the track may finish first.
/// It acts on whichever player is in front.
@Observable
final class SleepTimer {
    static let presetMinutes = [15, 30, 45, 60, 90]
    static let presetTrackCounts = [1, 2, 3, 5]
    private static let fadeSeconds = 10

    /// Seconds left in a countdown.
    private(set) var remaining: Int?
    /// The length the running countdown started with, for the home tab's
    /// preset chips.
    private(set) var countdown: Int?
    /// Stop once this many tracks finish (1 = the current one).
    private(set) var remainingTracks: Int?
    /// The countdown ended but the current track is allowed to finish.
    private(set) var waitingTrackEnd = false

    var isActive: Bool { remaining != nil || remainingTracks != nil || waitingTrackEnd }

    /// `播完本曲停` style suffix for the player's subtitle line.
    var statusText: String? {
        if let remaining { return "定时 \(Formatting.trackTime(ms: remaining * 1000))" }
        if waitingTrackEnd || remainingTracks == 1 { return "播完本曲停" }
        if let remainingTracks { return "还剩 \(remainingTracks) 曲停" }
        return nil
    }

    @ObservationIgnored weak var controller: (any SleepTarget)?
    private let prefs: PlayerPrefs
    @ObservationIgnored private var ticker: Task<Void, Never>?
    @ObservationIgnored private var fadeBase: Float?

    init(prefs: PlayerPrefs) {
        self.prefs = prefs
    }

    func start(seconds: Int) {
        reset()
        remaining = seconds
        countdown = seconds
        ticker = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(1))
                self?.tick()
            }
        }
    }

    func stopAfter(tracks: Int) {
        reset()
        remainingTracks = tracks
    }

    func cancel() {
        reset()
    }

    /// Called when a track finishes on its own; true means pause now.
    func trackCompleted() -> Bool {
        if waitingTrackEnd {
            reset()
            return true
        }
        guard let tracks = remainingTracks else { return false }
        if tracks > 1 {
            remainingTracks = tracks - 1
            return false
        }
        reset()
        return true
    }

    private func tick() {
        guard let controller, let current = remaining else { return }
        let next = current - 1
        guard next > 0 else {
            fire(controller)
            return
        }
        remaining = next
        if next <= Self.fadeSeconds && !prefs.sleepFinishCurrentTrack {
            let base = fadeBase ?? controller.volume
            fadeBase = base
            controller.volume = base * Float(next) / Float(Self.fadeSeconds)
        }
    }

    private func fire(_ controller: any SleepTarget) {
        let finishTrack = prefs.sleepFinishCurrentTrack && controller.isPlaying
        reset()
        if finishTrack {
            waitingTrackEnd = true
        } else {
            controller.pause()
        }
    }

    private func reset() {
        ticker?.cancel()
        ticker = nil
        remaining = nil
        countdown = nil
        remainingTracks = nil
        waitingTrackEnd = false
        if let fadeBase {
            controller?.volume = fadeBase
            self.fadeBase = nil
        }
    }
}
